opensoundscape/localization.py

```diff
@@ -256,3 +697,100 @@ def localize(
             return u0
         else:
             return u1
+
+
+def gillette_localize(
+    receivers=list,
+    delays=list,
+    temp=20,
+    m=[0],
+    exact=True,
+    summary=False,
+    confint=False,
+    alpha=0.05,
+    td_error=False,
+    total_td_error=False,
+):
+    """
+    Calculate the estimated location of a sound's source using the
+    algorithm laid out in Gillette and Silverman (2008)
+    Args:
+        receivers: A numpy array of coordinates for microphones used to
+        record the sound. The number of microphones needed should
+        be two more than the dimensions being localized in. The
+        first row will be treated as a reference point for the
+        algorithm.
+        tdoa: A list of time delays. Each entry should be the time
+        delay for the corresponding item in the receivers list
+        (i.e. the first item is the delay for the first receiver).
+        The first item in this list should be 0, with all other
+        entries centered around that.
+        temp: ambient temperature in Celsius. Defaults to 20.
+        exact: computes an exact solution if True, computes estimates
+        with uncertainty if false. Defaults to True
+        summary: displays a summary of the estimates if True. Defaults
+        to false.
+        confint: outputs confidence intervals for the estimated
+        coordinates if true. Defaults to false.
+        alpha: Determines confidence level of the confidence intervals.
+        Defaults to 0.05.
+        m: the index of the reference mic. Defaults to 0.
+        td_error: Computes the expected time delay from the estimated
+        source location, centered around the reference mic, for each
+        microphone.
+        total_td_error: Computes the euclidean norm of the errors
+        provided by td_error.
+    Returns:
+        an array with the estimated coordinates and the estimated
+        distance from the reference mic. (One reference mic and two
+        additional mics, this is a 2 item array containing an estima
+        -ted x coordinate and a distance.)
+    """
+    import opensoundscape.localization as loc
+    import statsmodels.api as sm
+
+    C = loc.calc_speed_of_sound(temperature=20)
+    # Compile know receiver locations and distance delays into an output vector
+    out_knowns = []
+    in_knowns = np.zeros(((len(receivers) - len(m)) * len(m), 2 + len(m)))
+    toa = np.array(delays)
+    r = 0
+    out_knowns = []
+    for k in range(len(m)):
+        tdoa = (
+            toa - toa[m[k]]
+        )  # Use the speed of sound to convert time delays to "distance delays"
+        diffs = []
+        for delay in tdoa:
+            diffs.append(float(delay * loc.calc_speed_of_sound(20)))
+        for i in range(len(receivers)):
+            if i in m:
+                continue
+            else:
+                w = diffs[i] ** 2
+                for j in range(len(receivers[i])):
```

Comment:
please explain and refer to equations in original text when possible. renaming variables with descriptive names should help a lot, plut lots of comments (applies to the remaining code below)